### PR TITLE
add extra header file to export

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
Export this header file for Horovod to throw and catch exception elegantly in MXNet.
Horovod PR: https://github.com/uber/horovod/pull/542

### CODE REVIEW

File: plugin/opencv/cv_api.cc [hunk 1]
```diff
@@ -26,9 +26,9 @@
 #include <dmlc/base.h>
 #include <mxnet/base.h>
 #include <mxnet/ndarray.h>
+#include <mxnet/c_api_common.h>
```
Comment (yuxihu) [inline plugin/opencv/cv_api.cc]: QQ: do we have guideline regarding the order of putting include files? Shall we follow alphabetic order of the filename? If so, please move this include to the right place here and all other places in this PR.

Comment (apeforest) [reply plugin/opencv/cv_api.cc]: I don't find a guideline for order. I was following the order of full path header followed by relative path header

File: include/mxnet/c_api_common.h [hunk 2]
```diff
@@ -22,8 +22,8 @@
  * \file c_api_error.h
  * \brief Error handling for C API.
  */
-#ifndef MXNET_C_API_C_API_COMMON_H_
-#define MXNET_C_API_C_API_COMMON_H_
+#ifndef MXNET_C_API_COMMON_H_
+#define MXNET_C_API_COMMON_H_
 
 #include <dmlc/base.h>
 #include <dmlc/logging.h>
```
Comment (eric-haibin-lin) [inline include/mxnet/c_api_common.h]: Do you need all these functions in the public header files, including `MXAPIThreadLocalEntry`, `CopyAttr`, etc?

Comment (eric-haibin-lin) [reply include/mxnet/c_api_common.h]: Do you mind pointing out how horovod will use these APIs?

**Comment by apeforest** [comment]
@eric-haibin-lin @szha The PR that requires this change is https://github.com/ctcyang/horovod/pull/19. The reason we need this header is that we are using `check_call` in Horovod to check the return status [here](https://github.com/ctcyang/horovod/pull/19/files#diff-6626ba53afb59b3f75cb017aa6259fd1R260). We need a way to catch the exception thrown from MXNet C API like [this](https://github.com/ctcyang/horovod/pull/19/files#diff-79b78539e15db1d75d73029490eed708R325). However, the `API_BEGIN()` and `API_END()` are defined in `c_api_common.h` which is the one I need to export.

File: include/mxnet/c_api_common.h (hunk 2 above)
Comment (apeforest) [reply include/mxnet/c_api_common.h]: @eric-haibin-lin I agree `MXAPIThreadLocalEntry`, `CopyAtt` are not need in error handling. I don't know the historical reason of adding them to this header file in the first place. My change was following the rationale of making least change to existing structure.

I could extract the error handling functions from `c_api_common.h` into `c_api_error.h` and only export `c_api_error.h` to include. Does that sound like a better approach to you?

**Comment by eric-haibin-lin** [comment]
I agree that MXNet should expose the c apis for error handling. Exposing API_BEGIN/API_END/API_END_HANDLE_ERROR sounds reasonable. We probably should prepend "MX_" in these macro. WDYT @apeforest @szha

**Comment by szha** [comment]
Let's limit the exposure to only what's required.

**Comment by apeforest** [comment]
@eric-haibin-lin @szha Per your suggestion, I have stripped out the error handling APIs into a separate header file and export that one only. The remaining code in MXNet modules remain unchanged. Please review this change again at your earliest convenience. Thanks a lot!

File: src/c_api/c_api_common.h [hunk 3]
```diff
@@ -29,37 +29,35 @@
...
-#define API_END_HANDLE_ERROR(Finalize) } catch(dmlc::Error &_except_) { Finalize; on_exit_api(); return MXAPIHandleException(_except_); } on_exit_api(); return 0; // NOLINT(*)
+#ifndef API_BEGIN
+#define API_BEGIN MX_API_BEGIN
+#endif
 
 /*!
- * \brief Set the last error message needed by C API
- * \param msg The error message to set.
+ * \brief every function starts with API_BEGIN();
+ * and finishes with API_END() or API_END_HANDLE_ERROR
```
Comment (yuxihu) [inline src/c_api/c_api_common.h]: can we consolidate the comments into one paragraph? It is kind of duplicate now. Maybe moving L46/L47 and L54-56 to after L39.

Comment (apeforest) [reply src/c_api/c_api_common.h]: updated

File: include/mxnet/c_api_error.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * \file c_api_error.h
+ * \brief Error handling for C API.
+ */
+#ifndef MXNET_C_API_ERROR_H_
+#define MXNET_C_API_ERROR_H_
+
+#include <dmlc/base.h>
+#include <dmlc/logging.h>
+#include <mxnet/c_api.h>
+#include <mxnet/base.h>
```
Comment (szha) [inline include/mxnet/c_api_error.h]: are all the headers needed?

Comment (apeforest) [reply include/mxnet/c_api_error.h]: good catch. removed unnecessary headers

**Pull request merged**